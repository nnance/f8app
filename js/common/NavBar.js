

import React from 'react';
import {
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
  Image,
  Dimensions,
} from 'react-native';

import { styles as commonStyles, NAV_BAR_HEIGHT, STATUS_BAR_HEIGHT } from './styles';

export const HEIGHT = NAV_BAR_HEIGHT + STATUS_BAR_HEIGHT;

const styles = StyleSheet.create({
  container: {
    paddingTop: STATUS_BAR_HEIGHT,
    height: NAV_BAR_HEIGHT,
    width: Dimensions.get('window').width,
    backgroundColor: 'white',
    justifyContent: 'space-between',
    flexDirection: 'row',
    alignItems: 'center',
  },
  titleMenu: {
    flex: 4,
    alignItems: 'center',
  },
  leftMenu: {
    flex: 1,
    paddingLeft: 10,
    alignItems: 'flex-start',
  },
  rightMenu: {
    flex: 1,
    paddingRight: 10,
    alignItems: 'flex-end',
  },
  navText: {
    fontSize: 22,
    fontWeight: 'bold',
  },
});

export default class NavBar extends React.Component {
  renderTitle() {
    return (
      <Text style={[styles.navText, this.props.titleTextStyle]}>{this.props.title}</Text>
    );
  }

  render() {
    return (
      <View style={[styles.container, this.props.containerStyle]}>
        <View style={[styles.leftMenu, this.props.leftMenuStyle]}>
          {
            this.props.renderLeftMenu ? this.props.renderLeftMenu() : null
          }
        </View>
        <View style={[styles.titleMenu, this.props.titleStyle]}>
          {
            this.props.renderTitle ? this.props.renderTitle() : this.renderTitle()
          }
        </View>
        <View style={[styles.rightMenu, this.props.rightMenuStyle]}>
          {
            this.props.renderRightMenu ? this.props.renderRightMenu() : null
          }
        </View>
      </View>
    );
  }
}

export class NavBarWithButton extends React.Component {
  constructor(...args) {
    super(...args);
    this.renderBackButton = this.renderBackButton.bind(this);
  }

  renderBackButton() {
    return (
      <TouchableOpacity onPress={this.props.onBackPress}>

              <View style={{ width: 40, height: 40, flex: 0, marginTop: 0, justifyContent: 'center', alignItems: 'center' }}>
          <Image style={commonStyles.navBarIcon} source={require('../assets/common/white-back-button.png')} />
        </View>
        </TouchableOpacity>
    );
  }

  render() {
    return (<NavBar
      renderLeftMenu={this.renderBackButton}
      {...this.props}
    />);
  }
}

export class NavBarWithPinkButton extends React.Component {
  constructor(...args) {
    super(...args);
    this.renderBackButton = this.renderBackButton.bind(this);
  }

  renderBackButton() {
    return (
      <TouchableOpacity onPress={this.props.onBackPress}>
        <View style={{ height: 30, width: 40, flex: 0, marginTop: 0, justifyContent: 'center', alignItems: 'center' }}>
          <Image style={commonStyles.navBarIcon} source={require('../assets/common/pink-back-button.png')} />
        </View>
        </TouchableOpacity>
    );
  }

  render() {
    return (<NavBar
      renderLeftMenu={this.renderBackButton}
      {...this.props}
    />);
  }
}
