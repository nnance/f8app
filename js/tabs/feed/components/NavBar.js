import React from 'react';

import { NavBarWithButton } from '../../../common/NavBar';

export { HEIGHT } from '../../../common/NavBar';

export default class NavBar extends React.Component {
  render() {
    return (
      <NavBarWithButton
        {...this.props}
      />
    );
  }
}
