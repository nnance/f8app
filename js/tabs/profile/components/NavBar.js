

import React from 'react';
import {
  StyleSheet,
  TouchableOpacity,
} from 'react-native';

import { NavBarWithPinkButton } from '../../../common/NavBar';

import { colors } from '../../../common/styles';

const styles = StyleSheet.create({
  container: {
    borderBottomWidth: 1,
    borderColor: colors.greyBorder,
  },
  backButton: {
    height: 20,
    resizeMode: 'contain',
  },
  titleText: {
    fontSize: 22,
    fontWeight: 'bold',
    color: colors.textPink,
  },
});

export default class NavBar extends React.Component {
  constructor(...args) {
    super(...args);
    this.renderRightMenu = this.renderRightMenu.bind(this);
  }

  renderRightMenu() {
    return (
      <TouchableOpacity style={styles.rightMenu} onPress={this.props.onRightPress}>
        {
          this.props.renderRightMenu ? this.props.renderRightMenu() : null
        }
      </TouchableOpacity>
    );
  }

  render() {
    return (
      <NavBarWithPinkButton
        {...this.props}
        containerStyle={styles.container}
        renderRightMenu={this.renderRightMenu}
        title={this.props.title}
        titleTextStyle={styles.titleText}
      />
    );
  }
}
