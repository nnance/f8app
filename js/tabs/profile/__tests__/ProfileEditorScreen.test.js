import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import {
  Switch,
} from 'react-native';

import TextInput from '../../../common/TextInput';

jest.mock('react-native-image-picker');

/* eslint import/first: off */
import ImagePicker from 'react-native-image-picker';

import NavBar from '../components/NavBar';
import ProfileEditorScreenComponent, { OKButton } from '../components/ProfileEditorScreen';

describe('ProfileEditorScreen', () => {
  const expectImage = { image: 'expect.png' };

  function api(name) {
    if (name === 'fail') {
      return Promise.reject(new Error('something wrong'));
    }
    return Promise.resolve();
  }

  it('OKButton', () => {
    const tree = renderer.create(<OKButton />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('call api changePublicProfile', async () => {
    let gResolve;
    const changePublicProfile = jest.fn(() => new Promise((resolve) => {
      gResolve = resolve;
    }));
    const wrapper = shallow(
      <ProfileEditorScreenComponent changePublicProfile={changePublicProfile} />,
    );
    wrapper.find(TextInput).simulate('changeText', 'AAA');
    expect(wrapper.state().savingProfile).toBe(false);
    const task = wrapper.find(NavBar).props().onRightPress();
    expect(wrapper.state().savingProfile).toBe(true);
    gResolve();
    await task;
    expect(wrapper.state().savingProfile).toBe(false);
    expect(changePublicProfile).toBeCalled();
    expect(changePublicProfile.mock.calls[0][0]).toBe('AAA');
  });

  it('call onBackPress if changed profile', async () => {
    const changePublicProfile = jest.fn(api);
    const onBackPress = jest.fn();
    const wrapper = shallow(
      <ProfileEditorScreenComponent
        changePublicProfile={changePublicProfile}
        onBackPress={onBackPress}
      />,
    );
    await wrapper.find(NavBar).props().onRightPress();
    expect(onBackPress).toBeCalled();
  });

  it('call api linkFacebook', async () => {
    let gResolve;
    const linkFacebook = jest.fn(() => new Promise((resolve) => {
      gResolve = resolve;
    }));
    const wrapper = shallow(
      <ProfileEditorScreenComponent linkFacebook={linkFacebook} facebookLinked={false} />,
    );
    expect(wrapper.state().linkingFacebook).toBe(false);
    const linkingPromise = wrapper.find(Switch).props().onValueChange();
    expect(linkFacebook).toBeCalled();
    expect(wrapper.state().linkingFacebook).toBe(true);
    gResolve();
    await linkingPromise;
    expect(wrapper.state().linkingFacebook).toBe(false);
  });

  it('call api unlinkFacebook', async () => {
    const unlinkFacebook = jest.fn(() => Promise.resolve());
    const wrapper = shallow(
      <ProfileEditorScreenComponent unlinkFacebook={unlinkFacebook} facebookLinked />,
    );
    await wrapper.find(Switch).props().onValueChange();
    expect(unlinkFacebook).toBeCalled();
  });

  it('open DatePickerDialog', async () => {
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    const expectDate = new Date(1993, 5, 12);
    const spy = jest.fn(() => Promise.resolve(expectDate));
    wrapper.instance().datePicker = {
      open: spy,
    };
    await wrapper.find('[name="birthDayInput"]').props().onPress();
    expect(spy).toBeCalled();
    expect(wrapper.state().birthDayDate.getTime()).toBe(expectDate.getTime());
  });

  it.skip('open SexPicker with male', async () => {
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    const Picker = require('react-native-picker').default;
    Picker.init.mockImplementationOnce(opts => opts.onPickerConfirm(['ชาย']));
    await wrapper.find('[name="sexInput"]').props().onPress();
    expect(Picker.init).toBeCalled();
    expect(wrapper.state().sex).toBe('M');
    Picker.init.mockClear();
  });

  it.skip('open SexPicker with female', async () => {
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    const Picker = require('react-native-picker').default;
    Picker.init.mockImplementationOnce(opts => opts.onPickerConfirm(['หญิง']));
    await wrapper.find('[name="sexInput"]').props().onPress();
    expect(Picker.init).toBeCalled();
    expect(wrapper.state().sex).toBe('F');
    Picker.init.mockClear();
  });

  it.skip('open SexPicker with unknow', async () => {
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    const Picker = require('react-native-picker').default;
    Picker.init.mockImplementationOnce(opts => opts.onPickerConfirm(['ไม่ระบุ']));
    await wrapper.find('[name="sexInput"]').props().onPress();
    expect(Picker.init).toBeCalled();
    expect(wrapper.state().sex).toBe(null);
    Picker.init.mockClear();
  });

  it('open openProfilePicker', async () => {
    ImagePicker.showImagePicker.mockImplementationOnce((opts, cb) => cb(expectImage));
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    await wrapper.find('[name="profileImageInput"]').props().onPress();
    expect(ImagePicker.showImagePicker).toBeCalled();
    expect(wrapper.state().changedProfilePicture).toBe(expectImage);
    ImagePicker.showImagePicker.mockClear();
  });

  it('open openProfileCoverPicker', async () => {
    ImagePicker.showImagePicker.mockImplementationOnce((opts, cb) => cb(expectImage));
    const wrapper = shallow(<ProfileEditorScreenComponent />);
    await wrapper.find('[name="coverImageInput"]').props().onPress();
    expect(ImagePicker.showImagePicker).toBeCalled();
    expect(wrapper.state().changedProfileCover).toBe(expectImage);
    ImagePicker.showImagePicker.mockClear();
  });
});
