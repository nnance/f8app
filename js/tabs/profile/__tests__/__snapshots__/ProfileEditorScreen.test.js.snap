exports[`ProfileEditorScreen OKButton 1`] = `
<Image
  source={1}
  style={
    Object {
      "height": 25,
      "resizeMode": "contain",
      "width": 25,
    }
  } />
`;
